ddl/column.go

```diff
@@ -693,17 +705,389 @@ func (w *worker) doModifyColumn(
+					tblInfo.Indices[i] = cIdx
+					break
+				}
+			}
+		}
+		changingCol.Name = colName
+		tblInfo.Indices = tblInfo.Indices[:len(tblInfo.Indices)-len(changingIdxs)]
+		// Adjust table column offset.
+		if err = adjustColumnInfoInModifyColumn(job, tblInfo, changingCol, oldCol, pos); err != nil {
+			// TODO: Do rollback.
+			return ver, errors.Trace(err)
+		}
+		updateChangingInfo(job, changingCol, changingIdxs, model.StatePublic)
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != changingCol.State)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// Finish this job.
+		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+		// TODO: Change column ID.
+		// asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfos: []*model.ColumnInfo{changingCol}})
+	default:
+		err = ErrInvalidDDLState.GenWithStackByArgs("column", changingCol.State)
+	}
+
+	return ver, errors.Trace(err)
+}
+
+func (w *worker) updatePhysicalTableRow(t table.PhysicalTable, oldColInfo, colInfo *model.ColumnInfo, reorgInfo *reorgInfo) error {
+	logutil.BgLogger().Info("[ddl] start to update table row", zap.String("job", reorgInfo.Job.String()), zap.String("reorgInfo", reorgInfo.String()))
+	return w.writePhysicalTableRecord(t.(table.PhysicalTable), typeUpdateColumnWorker, nil, oldColInfo, colInfo, reorgInfo)
+}
+
+// updateColumnAndIndexes handles the modify column reorganization state for a table.
+func (w *worker) updateColumnAndIndexes(t table.Table, oldCol, col *model.ColumnInfo, idxes []*model.IndexInfo, reorgInfo *reorgInfo) error {
+	// TODO: Consider rebuild ReorgInfo key to mDDLJobReorgKey_jobID_elementID(colID/idxID).
+	// TODO: Support partition tables.
+	err := w.updatePhysicalTableRow(t.(table.PhysicalTable), oldCol, col, reorgInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, idx := range idxes {
+		err = w.addTableIndex(t, idx, reorgInfo)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+type updateColumnWorker struct {
+	*backfillWorker
+	oldColInfo *model.ColumnInfo
+	newColInfo *model.ColumnInfo
+
+	// The following attributes are used to reduce memory allocation.
+	rowRecords []*rowRecord
+	rowDecoder *decoder.RowDecoder
+
+	rowMap             map[int64]types.Datum
+	idxKeyBufs         [][]byte
+	batchCheckKeys     []kv.Key
+	distinctCheckFlags []bool
+}
+
+func newUpdateColumnWorker(sessCtx sessionctx.Context, worker *worker, id int, t table.PhysicalTable, oldCol, newCol *model.ColumnInfo, decodeColMap map[int64]decoder.Column) *updateColumnWorker {
+	rowDecoder := decoder.NewRowDecoder(t, t.WritableCols(), decodeColMap)
+	return &updateColumnWorker{
+		backfillWorker: newBackfillWorker(sessCtx, worker, id, t),
+		oldColInfo:     oldCol,
+		newColInfo:     newCol,
+		rowDecoder:     rowDecoder,
+		rowMap:         make(map[int64]types.Datum, len(decodeColMap)),
+	}
+}
+
+type rowRecord struct {
+	handle kv.Handle
```

Comment:
It seems only the last accessed handle is used. Can we remove this field to reduce memory consumption?

Reply:
`handle` is used for `getNextHandle`.